### PR TITLE
MULE-10568 Support non-blocking message sources

### PR DESCRIPTION
**Pull request by dfeist**
This is the cleanest approach of about 3 or 4 different options I tried that does not impact the whole code base. I still need to ensure all tests pass.

### CODE REVIEW

File: core/src/main/java/org/mule/runtime/core/api/source/AsyncMessageSource.java [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+ * @since 4.0
+ */
+public interface AsyncMessageSource extends MessageSource {
+
+  /**
+   * Set the {@link AsyncProcessor} listener on a message source which will be invoked when a message is received or generated.
+   *
+   * @param listener the listener.
+   */
+  void setAsyncListener(AsyncProcessor listener);
```
Comment (marianogonzalez) [inline core/src/main/java/org/mule/runtime/core/api/source/AsyncMessageSource.java]: setAsyncProcessor

File: core/src/main/java/org/mule/runtime/core/construct/AbstractPipeline.java [hunk 2]
```diff
@@ -416,6 +428,10 @@ protected void doDispose() {
     super.doDispose();
   }
 
+  protected boolean useBlockingCodePath() {
```
Comment (marianogonzalez) [inline core/src/main/java/org/mule/runtime/core/construct/AbstractPipeline.java]: isBlocking() or shouldBlock()

File: core/src/main/java/org/mule/runtime/core/execution/ModuleFlowProcessingPhase.java [hunk 3]
```diff
@@ -64,118 +66,182 @@ public boolean supportsTemplate(MessageProcessTemplate messageProcessTemplate) {
...
+      Consumer<MessagingException> errorConsumer =
+          getErrorConsumer(messageSource, template.getFailedExecutionResponseParametersFunction(),
+                           messageProcessContext, template, phaseResultNotifier);
+
+      Event templateEvent = Event
+          .builder(create(messageProcessContext.getFlowConstruct(), sourceIdentifier.getNamespace()))
+          .message((InternalMessage) template.getMessage()).build();
+
+      // TODO MULE-11167 Policies should be non blocking
+      if (System.getProperty("enableSourcePolicies") == null) {
```
Comment (marianogonzalez) [inline core/src/main/java/org/mule/runtime/core/execution/ModuleFlowProcessingPhase.java]: constant

File: core/src/main/java/org/mule/runtime/core/execution/ModuleFlowProcessingPhase.java [hunk 4]
```diff
@@ -64,118 +66,182 @@ public boolean supportsTemplate(MessageProcessTemplate messageProcessTemplate) {
...
+
+        just(templateEvent)
+            .doOnNext(request -> fireNotification(messageProcessContext.getMessageSource(), request,
+                                                  messageProcessContext.getFlowConstruct(),
+                                                  MESSAGE_RECEIVED))
+            .then(request -> from(template.routeEventAsync(request)))
+            .doOnSuccess(getSuccessConsumer(messageSource, templateEvent, exceptionHandler, errorConsumer,
```
Comment (marianogonzalez) [inline core/src/main/java/org/mule/runtime/core/execution/ModuleFlowProcessingPhase.java]: methods with so many arguments are usually a bad sign

File: modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/runtime/source/ExtensionMessageSource.java [hunk 5]
```diff
@@ -295,9 +297,13 @@ public Object getWorkOwner() {
 
   @Override
   public void setListener(Processor listener) {
-    messageProcessor = listener;
+    // Ignore, this source only needs AsyncProcessor listener
```
Comment (marianogonzalez) [inline modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/runtime/source/ExtensionMessageSource.java]: this should not ignore. This method should ideally not exist or at most throw UnsupportedOperationException

File: core/src/main/java/org/mule/runtime/core/execution/ModuleFlowProcessingPhase.java (hunk 4 above)
Comment (dfeist) [inline core/src/main/java/org/mule/runtime/core/execution/ModuleFlowProcessingPhase.java]: I know, this class is already in desperate need to refactor and unit tests. I gave this feedback already in a previous PR. This here is temporary until we sort out non-blocking polcies at which point this whole class needs a clean up and addition of unit tess.

File: core/src/main/java/org/mule/runtime/core/execution/ModuleFlowProcessingPhase.java [hunk 6]
```diff
@@ -51,6 +51,7 @@
 public class ModuleFlowProcessingPhase
     extends NotificationFiringProcessingPhase<ModuleFlowProcessingPhaseTemplate> {
 
+  public static final String ENABLE_SOURCE_POLICIES_SYSTEM_PROPERTY = "enableSourcePolicies";
```
Comment (marianogonzalez) [inline core/src/main/java/org/mule/runtime/core/execution/ModuleFlowProcessingPhase.java]: isn't MuleProperties the place where we group all constants?

File: core/src/main/java/org/mule/runtime/core/construct/AbstractPipeline.java (hunk 2 above)
Comment (dfeist) [inline core/src/main/java/org/mule/runtime/core/construct/AbstractPipeline.java]: `isBlocking` for sure if this was a field being exposed via a javabeans getter/setter.  But it is an internal behavioural method and `use` is just as good as `should` or `shouldUse` for this.

File: core/src/main/java/org/mule/runtime/core/execution/ModuleFlowProcessingPhase.java (hunk 6 above)
Comment (dfeist) [inline core/src/main/java/org/mule/runtime/core/execution/ModuleFlowProcessingPhase.java]: This should be gone in max 2 weeks when @pablolagreca is back. No point putting something temporary in MuleProperties. This is only hear to allow testing with policies in the next two weeks if required.

File: core-tests/src/test/java/org/mule/runtime/core/construct/AbstractFlowConstructTestCase.java [hunk 7]
```diff
@@ -24,16 +23,12 @@
 
   public static class DirectInboundMessageSource implements MessageSource {
 
-    private Processor listener;
+    Processor listener;
```
Comment (elrodro83) [inline core-tests/src/test/java/org/mule/runtime/core/construct/AbstractFlowConstructTestCase.java]: provide a getter and make private

File: core-tests/src/test/java/org/mule/runtime/core/construct/FlowTestCase.java [hunk 8]
```diff
@@ -45,20 +46,36 @@
...
   private SensingNullMessageProcessor sensingMessageProcessor;
+  private Mode mode;
+
+  public FlowTestCase(Mode mode) {
+    this.mode = mode;
+  }
+
+  @Parameters
+  public static Collection<Object[]> parameters() {
+    return asList(new Object[][] {{Mode.BLOCKING}, {Mode.ASYNC}});
```
Comment (elrodro83) [inline core-tests/src/test/java/org/mule/runtime/core/construct/FlowTestCase.java]: you may use a Function and avoid having an enum

File: core-tests/src/test/java/org/mule/runtime/core/construct/FlowTestCase.java [hunk 9]
```diff
@@ -253,4 +270,16 @@ public void testFailStartingMessageSourceOnLifecycleShouldStopStartedPipelinePro
     verify((Startable) mockMessageSource, times(1)).start();
     verify((Stoppable) mockMessageSource, times(1)).stop();
   }
+
+  private Event triggerSource(Event event) throws MuleException {
+    if (mode == Mode.ASYNC) {
+      return Mono.just(event).transform(directInboundMessageSource.listener).block();
```
Comment (elrodro83) [inline core-tests/src/test/java/org/mule/runtime/core/construct/FlowTestCase.java]: static import

File: core/src/main/java/org/mule/runtime/core/construct/AbstractPipeline.java [hunk 10]
```diff
@@ -268,6 +268,14 @@ public Event process(final Event event) throws MuleException {
     createFlowMap();
   }
 
+  Function<Publisher<Event>, Publisher<Event>> processFlowFunction() {
+    return mono -> Mono.from(mono).transform(pipeline)
```
Comment (elrodro83) [inline core/src/main/java/org/mule/runtime/core/construct/AbstractPipeline.java]: static import

File: core/src/main/java/org/mule/runtime/core/execution/ModuleFlowProcessingPhase.java (hunk 6 above)
Comment (elrodro83) [inline core/src/main/java/org/mule/runtime/core/execution/ModuleFlowProcessingPhase.java]: add a TODO here then

File: modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/runtime/source/ExtensionMessageSource.java [hunk 11]
```diff
@@ -295,10 +295,9 @@ public Object getWorkOwner() {
 
   @Override
   public void setListener(Processor listener) {
-    messageProcessor = listener;
+    this.messageProcessor = listener;
```
Comment (elrodro83) [inline modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/runtime/source/ExtensionMessageSource.java]: the attribute should be renamed it seems

File: core-tests/src/test/java/org/mule/runtime/core/construct/AbstractFlowConstructTestCase.java [hunk 12]
```diff
@@ -21,14 +21,19 @@
...
 
-    Processor listener;
+    private Processor listener;
 
     public void setListener(Processor listener) {
       this.listener = listener;
     }
 
+    public Processor getListener()
+    {
```
Comment (elrodro83) [inline core-tests/src/test/java/org/mule/runtime/core/construct/AbstractFlowConstructTestCase.java]: format

**Pull request merged**